Handle text features

HLasse opened the pull request:
Adds the `TextSpec` dataclass which allows us to handle text features. Further, adds text embeddings function for huggingface, sentence-transformers, and sklearn models. 

- [ ] I have assigned ranges (e.g. `>=0.1, <0.2`) to all new dependencies (allows dependabot to keep dependency ranges wide for better compatability)

Fixes #97 .

## TODO
- [x] Write `TextSpec`
- [x] Make `_flatten_temporal_values_to_df` handle multiple values
- [x] Update documentation to include TextSpec
- [x] Add tutorial showing how to create embedding functions and use TextSpec
- [ ] (Make cache work for text features/features that output multiple columns)
- [x] (write Huggingface embedding function) 

## Notes for reviewers
Reviewers can skip X, but should pay attention to Y.

MartinBernstorff commented on src/timeseriesflattener/flattened_dataset.py:
```diff
@@ -303,6 +304,24 @@ def _drop_records_outside_interval_days(
             axis=1,
         )
 
+    @staticmethod
+    def _embed_text_values(
+        df: pd.DataFrame,
+        embedding_fn: Callable,
+        dim_reduction_fn: Optional[Callable],
+    ) -> pd.DataFrame:
+        """Embeds text values using the embedding_fn and optionally reduces the
+        dimensionality using dim_reduction_fn. Assumes that the text is stored
+        in the column "value"."""
+        embedding = embedding_fn(df["value"])
+        if dim_reduction_fn:
+            embedding = dim_reduction_fn(embedding)
+
+        df = df.drop("value", axis=1)
+        # make multiindex with embedding as 'value'
+        df = pd.concat([df, embedding], axis=1, keys=["df", "value"])
+        return df
```
Since this is a static method, maybe it should be moved to a separate module outside of timeseriesflattener?

This goes for quite a few of the methods, but this is distinctly different from flattening of timeseries, so an easy candidate to start with :-)

MartinBernstorff commented on src/timeseriesflattener/flattened_dataset.py:
```diff
@@ -303,6 +304,24 @@ def _drop_records_outside_interval_days(
+    @staticmethod
+    def _embed_text_values(
+        df: pd.DataFrame,
+        embedding_fn: Callable,
```
Ideally, we would specify this as an abstract class to make the interface more explicit, making it easier to add new embedding functions.

HLasse replied:
I've specified the format in the docstring now (quite simple: pd.Series with text in, pd.DataFrame out), and added some example embedding functions. I think ABC would be a bit overkill in this case (similar to resolve_multiple_fns)

MartinBernstorff reviewed (commented):
This looks friggin' cool, Lasse!

Super happy we have tests here - both the ones you wrote, and for the core functionality - because we're touching quite a bit of the core logic.

I love the code you wrote, but worry that the `flattened_dataset` class gets too unwieldy with this many lines - it was already borderline, and that's my fault. 

One suggestion is to decompose it, for example moving all the multiindex static methods into a separate class, as well as moving text handling into their own class as well. Does that sound good to you?

HLasse commented:
Thanks for the great feedback, Martin! And yep, the tests make working in and tweaking the library so much better!!
I've refactored the multiindex things out, moved the text embedding fn and added some examples. 

If you're happy with the current setup, I'll add some docs and write a short tutorial :)

MartinBernstorff commented:
Super happy with the current setup, and agreed re: legibility - I think this is the correct trade-off 👍

HLasse commented:
Alright, should be good to go!
Main new things:
1. Added functions for using huggingface, sentence-transformers, and sklearn models as embedding functions.
2. Added tutorial showing how to use these functions as well as how to write your own
3. Refactored `_flatten_temporal_values_to_df` so the prediction times that do not have a value are not added to the dataframe until resolve_multiple and embedding has been done. This should speed up the computation a bit (since there are vastly fewer rows) and greatly simplifies the text functions (as they don't have to deal with NaNs). I can't see why it should happen before resolve_multiple as NaNs are not used in any of them, but let me know if I've missed something!

Tests seem to be failing due to a poetry thing - @MartinBernstorff, got any clues what to do?

HLasse commented:
Ah, I see - that's probably it. Strongly in favour of migrating to pip, would be amazing if you could take a look at it tomorrow!

HLasse commented:
Alright, so it seems that the tests that used models from Huggingface caused the action to hang. I'm investigating possible solutions to the problem, but for now I think the most practical solution is to decorate them with a pytest marker so they are run locally but not on github (e.g. something like this https://stackoverflow.com/questions/47559524/pytest-how-to-skip-tests-unless-you-declare-an-option-flag)

MartinBernstorff commented:
> Alright, so it seems that the tests that used models from Huggingface caused the action to hang. I'm investigating possible solutions to the problem, but for now I think the most practical solution is to decorate them with a pytest marker so they are run locally but not on github (e.g. something like this https://stackoverflow.com/questions/47559524/pytest-how-to-skip-tests-unless-you-declare-an-option-flag).
> 
> Also seems to be a problem for Huggingface themselves: [huggingface/transformers#11408](https://github.com/huggingface/transformers/issues/11408)

Sounds great re: the decorator. Giving this a look now.

MartinBernstorff commented on src/timeseriesflattener/text_embedding_functions.py:
```diff
@@ -0,0 +1,95 @@
+    embedding_fn: Callable,
+    embedding_fn_kwargs: Optional[dict] = None,
+) -> pd.DataFrame:
+    """Embeds text values using the embedding_fn and optionally reduces the
+    dimensionality using dim_reduction_fn. Stores the embedding as a multi-
+    index column called 'value'.
+
```
You removed the dimensionality reduction functionality completely? :-)

HLasse replied:
Yes, after giving it some more though, it makes much more sense to handle during model training. It requires a model that's fit to the embedding, so it's way more convenient, fast, (and good hyperparam) to do afterwards

MartinBernstorff replied:
Ah, yes, it's actually a post-split preprocessing step! Good catch!

MartinBernstorff commented on src/timeseriesflattener/flattened_dataset.py:
```diff
@@ -402,124 +380,54 @@ def _flatten_temporal_values_to_df(  # noqa pylint: disable=too-many-locals
             output_spec=output_spec,
         )
-        df = TimeseriesFlattener._add_fallback_to_value_cols(
+        df = MultiIndexHandler.add_fallback_to_value_cols(
```
```suggestion
        df = MultiIndexHandler.replace_na_in_value_col_with_fallback(
```

I assume this is what the function does? :-)

MartinBernstorff commented on src/timeseriesflattener/flattened_dataset.py:
```diff
@@ -402,124 +380,54 @@ def _flatten_temporal_values_to_df(  # noqa pylint: disable=too-many-locals
-        feature_col_name_mapping = dict(zip(feature_names, col_names))
-        # level=1 means that the column names are in the second level of the multiindex
-        df = df.rename(columns=feature_col_name_mapping, level=1)
+        """Checks whether 'value' is a column in df, and if not, renames the column"""
```
```suggestion
        """Checks whether 'value' is a column in df, and if not, renames the input column"""
```

MartinBernstorff commented on src/timeseriesflattener/multi_index_handling.py:
```diff
@@ -0,0 +1,98 @@
+            df (pd.DataFrame): Dataframe with value column
+            output_spec (TemporalSpec): Output specification
+        """
+        if "value" in df.columns and isinstance(df["value"], pd.DataFrame):
+            df["value"] = df["value"].fillna(output_spec.fallback)
+        else:
+            df[output_spec.get_col_str()] = df[output_spec.get_col_str()].fillna(
+                output_spec.fallback,
+            )
+        return df
```
Can we have cases where the value col is a multiindex but isn't called "value"? If not, I'm fine with this 👍

HLasse replied:
Nope, that's handled in `TimeseriesFlattener.rename_input_col_to_value` (necesarry to have the `input_col_name_override` work for temporal specs)

MartinBernstorff commented on src/timeseriesflattener/multi_index_handling.py:
```diff
@@ -0,0 +1,98 @@
+        return df
+
+    @staticmethod
+    def add_fallback_to_value_cols(
```
```suggestion
    def replace_na_in_spec_col_with_fallback(
```

Maybe even more specific, since it can in principle take any specification.

MartinBernstorff commented on src/timeseriesflattener/multi_index_handling.py:
```diff
@@ -0,0 +1,98 @@
+        Args:
+            df (pd.DataFrame): Dataframe to get value column name from.
+            output_spec (TemporalSpec): Output specification"""
+        if isinstance(df.columns, pd.MultiIndex):
+            return df["value"].columns.tolist()
+        else:
+            return [output_spec.get_col_str()]
```
The cohesion of this function seems somewhat low - it's two very different things. E.g. it has two arguments, but only the "if" uses the first argument and only the "else" uses the second argument. 

Perhaps it shouldn't be a function, but split up?

HLasse replied:
I see the argument, but obtaining the name(s) of the columns requires a bit different handling (as you see) depending on whether it's multiindex or not. This seemed to me the most elegant way of keeping the code clean and readable in the main module. I guess you could argue that the cohesion within the control flow in the function is low, but it's necesarry to handle the multiindex and single column cases. 
You are suggesting to not use a function but have the control flow in `TimeseriesFlattener._flatten_temporal_values_to_df` instead?

MartinBernstorff replied:
Hmm, I see your point. 

I still think it's kind of weird that we have two arguments, where it'd only ever make sense to pass one of them. I.e. they're both optional, and they're kind of unrelated. 

My quip maybe originates from this being called `MultiIndexHandler`, even though ~1/3 of the logic is unrelated to multiindexing. Perhaps it should be something like `ColumnHandler?`

In the spirit of pragmatism, I suggest calling the `ColumnHandler` and then merging :-)

MartinBernstorff commented on src/timeseriesflattener/testing/load_synth_data.py:
```diff
@@ -122,4 +122,4 @@ def load_synth_text(
         pd.DataFrame
     """
-    return load_raw_test_csv("synth_txt_data.csv", n_rows=n_rows)
+    return load_raw_test_csv("synth_text_data_matched.csv", n_rows=n_rows)
```
What does matched mean here? :-)

HLasse replied:
Good question! That the values in the entity_id column match those from the synthetic prediction times. Can rename to "synth_text_data.csv" instead and remove the old file

MartinBernstorff commented on src/timeseriesflattener/text_embedding_functions.py:
```diff
@@ -0,0 +1,95 @@
+    """Embeds the text data using a huggingface model. To use this in timeseriesflattener,
+    supply the model_name as an embedding_fn_kwargs argument to TextPredictorSpec.
+    For example:
+    `embedding_fn_kwargs={"model_name": "bert-base-uncased"}`
```
Love this ❤️

MartinBernstorff commented on tests/test_timeseriesflattener/test_embedding_functions.py:
```diff
@@ -3,11 +3,36 @@
     assert embedding_df.shape == (df.shape[0], 10)
+
+
+# def test_huggingface_embedding(synth_text_data):
+#     """Test that the huggingface embedding function works as expected"""
+#     df = synth_text_data.dropna(subset="text")
+#     df = df.head(5)
+#     embedding_df = huggingface_embedding(
+#         df["text"],
+#         model_name="sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2",
+#     )
+#     assert embedding_df.shape == (df.shape[0], 384)
+
+
+# def test_sentence_transformer_embedding(synth_text_data):
+#     """Test that the sentence-transformer embedding function works as expected"""
+#     df = synth_text_data.dropna(subset="text")
+#     df = df.head(5)
+#     embedding_df = sentence_transformers_embedding(
+#         df["text"],
+#         model_name="sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2",
+#     )
+#     assert embedding_df.shape == (df.shape[0], 384)
```
Decorating this with a pytest fixture seems like a good solution 👍

MartinBernstorff commented on tests/test_timeseriesflattener/test_flattened_dataset/test_add_values.py:
```diff
@@ -85,13 +82,12 @@ def test_text_predictor():
             feature_name="text_value",
         ),
-        expected_values=[[-15.409306032473708, -0.536602322891607]],
+        expected_values=[[0] * 10],
```
```suggestion
        expected_values=[[0] * 10],
```

Are these reasonable values?

HLasse replied:
Yup, the count vectorizer is fit the to synthetic data (which goes something "the patient has ..."). Neither "hello" nor "world" are very are very prevalent in them :)

MartinBernstorff reviewed (changes requested):
Really happy with the refactor, Lasse! Minor comments, think this is very close to merging 👍

MartinBernstorff commented:
In an ideal world, getting a small performance benchmark would be amazing. I have no idea how fast this'll be. That said, we'll get it when we first apply it.

I know this is super late in the game, but do we want to do the embedding in timeseriesflattener? Or should we just concatenate the text in the lookbehind, and then do embedding after flattening?

HLasse commented:
Alright, I think it should be good to go! 

Yep, let's see when we use it! Going to vastly differ by model I'm sure.

Re: embedding in timeseriesflattener. Yes, I definitely think we want it here. Embedding takes time, but only has to be run once. If you want multiple embeddings it would be easy to create, and then subset during training instead. Seems like a much better trade-off to spend 30 minutes extra on feature generation compared to 30 mins extra per model training - unless I'm misunderstanding? :)

MartinBernstorff commented:
> Re: embedding in timeseriesflattener. Yes, I definitely think we want it here. Embedding takes time, but only has to be run once. If you want multiple embeddings it would be easy to create, and then subset during training instead. Seems like a much better trade-off to spend 30 minutes extra on feature generation compared to 30 mins extra per model training - unless I'm misunderstanding? :)

I agree re: running once, but I wonder whether `timeseriesflattener` is the right place - maybe `psycop-feature-generation` would be better suited?

That said, this is basically done, and I could see users of timeseriesflattener wanting to use text-features. Let's keep it here for now, even though it's not 100% cohesive :-)

Great work! Merging 👍

MartinBernstorff commented on src/timeseriesflattener/multi_index_handling.py:
```diff
@@ -0,0 +1,98 @@
+        return df
+
+    @staticmethod
+    def get_value_col_str_name(
+        df: pd.DataFrame,
+        output_spec=TemporalSpec,
+    ) -> List[str]:
```
```suggestion
    def get_value_col_str_name(
        df: Optional[pd.DataFrame] = None,
        output_spec: Optional[TemporalSpec] = None,
    ) -> List[str]:
```

We might want a guard for error handling here, e.g. what happens if neither are passed?

MartinBernstorff commented:
Just three minor points in the unresolved convos, then LGTM!

HLasse commented:
I see what you mean. Text is a weird case (compared to continuous/binary features) so I see the argument for having it done after. That being said, I think it's more pragmatic to keep it in timeseriesflattener as, as you say, I think people will use it and appreciate it being in the same flow as the rest of their feature generation. 

The option of "just" concatenating is still there if people want to only use that though - would just require specifying the text loader as a regular `PredictorSpec` and using "concatenate" as resolve multiple.
